_posts/2019-12-05-testowanie-komponentow-z-inputami-i-outputami.md

```diff
@@ -0,0 +1,198 @@
+  let component: CalculatorInputFieldComponent;
+  let fixture: ComponentFixture<CalculatorInputFieldComponent>;
+  let input: DebugElement;
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [CalculatorInputFieldComponent]
+    }).compileComponents();
+
+    fixture = TestBed.createComponent(CalculatorInputFieldComponent);
+    component = fixture.componentInstance;
+    input = fixture.debugElement.query(By.css('input'));
+  }));
+
+  it('should emit correct value on input change', () => {
+    component.fieldValue.subscribe((value) => expect(value).toBe('7'));
+    input.nativeElement.value = 7;
+    document.querySelector('input').dispatchEvent(new KeyboardEvent('keyup'));
+  });
+});
+```
+
+Czyli standardowo, zaczynamy od przygotowania naszego modułu testowego i uchwycenia elementu, który chcemy przetestować. Warto tutaj zaznaczyć, że istnieją inne sposoby na zainicjowanie komponentu niż `TestBed`, wszystko zależy od tego co, i w jaki sposób, będziemy testować. Na przykład kiedy nie potrzebujemy testować elementów z DOM możemy zostać przy zwykłym inicjalizowaniu komponentów. Więcej szczegółów można znaleźć w [dokumentacji od Angulara](https://angular.io/guide/testing#component-test-basics).
+
+Test zaczynamy od nasłuchiwania na naszą zmienną `fieldValue` po której spodziewamy się, że przekaże nam wartość wpisaną w pole. Następnie uzupełniamy wartość i wywołujemy event, pozwalający wywołać metodę nasłuchującą na zmiany w naszym polu. Test zakończy się, kiedy zostanie wywołana metoda `expect` sprawdzająca, czy wpisana przez nas wartość zgadza się z oczekiwaniami. Test może się również zakończyć niepowodzeniem, kiedy po upływie danego czasu (domyślna konfiguracja wskazuje na 5s) żaden `expect` nie zostanie wywołany.
+
+Na koniec możemy zebrać naszą wiedzę, i przetestować komponent zawierający wejścia, jak i wyjścia. Poniższy kod przedstawia komponent, który dodaje wskazane liczby oraz opcjonalnie potrafi wysłać dane o wyniku do komponentu 'wyżej'.
+
+```typescript
+@Component({
+  selector: 'calculator-result',
+  template: `
+      <calculator-result-presentation [calculationResult]="getCalculationResult()"></calculator-result-presentation>
+      <button (click)="saveValue()">Zapisz wartość</button>
+  `,
+})
+export class CalculatorResultComponent {
+  @Input() values: Array<number>;
+  @Input() text: string;
+  @Output() savedValue = new EventEmitter<number>();
+
+  getResult(): number {
+    return Number(this.values[0]) + Number(this.values[1]);
+  }
+
+  getCalculationResult(): string {
+    return `${this.text} ${this.getResult()}`;
+  }
+
+  saveValue() {
+    this.savedValue.emit(this.getResult());
+  }
+}
+```
+
+Możemy zauważyć, że nasz komponent potrzebuje innego komponentu (`calculator-result-presentation`), któremu przekazuje wynik, jednakże w naszych testach nie będzie miało to dużego znaczenia. Należy tylko pamiętać o jego deklaracji przy tworzeniu `TestBed`.
```

Comment:
przenisósłym te część do drugiego zdania:
, jednakże w naszych testach nie będzie miało to dużego znaczenia. Należy tylko pamiętać o jego deklaracji przy tworzeniu `TestBed`. -> . 
W testach nie będzie miało to dużego znaczenia, należy jednak pamiętać o jego deklaracji przy tworzeniu `TestBed`.